pkg/operator/csidriveroperator/crcontroller.go

```diff
@@ -0,0 +1,266 @@
+package csidriveroperator
+
+import (
+	"context"
+	"fmt"
+	"strings"
+	"time"
+
+	operatorapi "github.com/openshift/api/operator/v1"
+	opclient "github.com/openshift/client-go/operator/clientset/versioned"
+	oplisters "github.com/openshift/client-go/operator/listers/operator/v1"
+	"github.com/openshift/cluster-storage-operator/pkg/csoclients"
+	"github.com/openshift/cluster-storage-operator/pkg/generated"
+	"github.com/openshift/cluster-storage-operator/pkg/operator/csidriveroperator/csioperatorclient"
+	"github.com/openshift/library-go/pkg/controller/factory"
+	"github.com/openshift/library-go/pkg/operator/events"
+	"github.com/openshift/library-go/pkg/operator/resource/resourceapply"
+	"github.com/openshift/library-go/pkg/operator/resource/resourcehelper"
+	"github.com/openshift/library-go/pkg/operator/resource/resourcemerge"
+	"github.com/openshift/library-go/pkg/operator/status"
+	"github.com/openshift/library-go/pkg/operator/v1helpers"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/runtime/serializer"
+	"k8s.io/apimachinery/pkg/util/errors"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/klog/v2"
+)
+
+// This CSIDriverOperatorCRController installs and syncs CSI driver operator CR. It monitors the
+// CR status and merges all its conditions to the CSO CR.
+// It produces following Conditions:
+// <CSI driver name>CSIDriverOperatorControllerDegraded on error
```

Comment:
This controller actually produces 2 types of degraded conditions:

* `WithSyncDegradedOnError()` will produce a `c.name + "Degraded"` condition

* `Sync()` will produce a `c.name + csiDriverControllerName + "Degraded"` condition

IMO ideally we should have only 1.

Reply:
The controller should produce two Degraded conditions because of `WithSyncDegradedOnError` - it clears its condition when sync() returns nil. And we still want copied Degraded conditions from the CR. 

I added an explicit comment about that + separated their prefixes
```
// <CSI driver name>CSIDriverOperatorDegraded on error
// <CSI driver name>CSIDriverOperatorCRDegraded - copied from *Degraded conditions from CR.
// <CSI driver name>CSIDriverOperatorCRAvailable - copied from *Available conditions from CR.
// <CSI driver name>CSIDriverOperatorCRProgressing - copied from *Progressing conditions from CR.
```

(as a separate commit, because this one is a bigger change)